### DISCUSSION TITLE
How does CVXPY transform a high-level quadratic programming into the standard form?

### DISCUSSION THREAD

**Discussion by Jushan-Randy-Chen** [cvxpy/cvxpy#2142] [Q&A] [answered]
I am trying to solve an MPC problem with multiple agents, and I want to use the OSQP solver interface in CVXPY. I initally tried using the original [OSQP package](https://osqp.org/docs/solver/index.html) directly to solve my problem, but the standard form in OSQP is given as the following:

![image](https://github.com/cvxpy/cvxpy/assets/56526142/9da9231b-0db1-439c-9961-3c86299b5c93)

This is not a very convenient formulation since I want collision avoidance cost to be added to the objective function. For example, I add a penalty term that is a function of the distance between agent `i` and agent `j` at all time steps, but I can't figure out a straight forward way to convert that into the matrix form.

Therefore, I wonder how exactly does CVXPY resolve such an issue behind the scene? It seems very easy to use CVXPY modeling language because I simply need to use the `cvxpy.sum_squares()` function. Since CVXPY provides an interface to `OSQP`, I guess it must also transform the problem to a sturcture which `OSQP` recognizes.

Any help is greatly appreciated!

**Comment by phschiele** [comment, marked as answer]
Hi @RandyChen233!

Your question precisely emphasises one main strength of CVXPY: As a user, you can use all atoms, without necessarily caring how to transform them to the standard form.

Nevertheless, it is great to understand what is going on behind the scenes, so I want to give you a few tools that might help you deepen your understanding.

Lets start off with a toy problem:

```py
import cvxpy as cp
n = 2
x = cp.Variable(n, nonneg=True)
y = cp.Variable(n, nonneg=True)
objective = cp.sum_squares(x + y)
problem = cp.Problem(cp.Minimize(objective))
```

Let's solve the problem using `problem.solve(verbose=True)` and look at the output.
In there, you should see this line:
> Reduction chain: CvxAttr2Constr -> Qp2SymbolicQp -> QpMatrixStuffing -> OSQP

These are the reductions, that take in a problem, and at each step return an equivalent problem that is close to the standard form you posted.

Let's go left to right:
The first item in the chain transforms attributes into actual constraints. For example, our `nonneg=True` attribute simply becomes the constraint `x>=0`. Note that we did not have any explicit constraints in the problem formulation, but now we already know that we can expect there to be constraints in the transformed problem.

Now we see `Qp2SymbolicQp`, which contains the transformations that represent an atom equivalently as a quadratic program close to the standard form. If you look into `qp2symbolic_qp.py` you find a reference to the `qp_canon_methods`. These are exactly the transformation functions for the atoms. If you look at the definition in `sum_squares`, you see that it is actually defined as `quad_over_lin(expr, 1)`, which makes sense as the latter is $(\sum_{ij}X^2_{ij})/y$, and setting $y$ equal to 1 is exactly the sum of the squares. You can now look up the canonicalization of  `quad_over_lin`, which is in `quad_over_lin_canon.py`.
You find that in our case, it is basically just setting P to `eye(affine_expr.size)/y.value`. You'll see later, however, that in our case, we do not just have a variable that we want to find the sum of the squares, but a sum of variables! As you can see, CVXPY automatically introduces a new variable `t`, and adds the constraint `t==x+y`.
Note: If we would have a conic program instead of a QP, this step would be the `dcp2cone` reduction instead.
Another note: you have to be careful which `quad_over_lin_canon.py` file you look at, as they exist for the QP and cone pathway (the latter is used when y is not constant). I am referring to `cvxpy/reductions/qp2quad_form/canonicalizers/quad_over_lin_canon.py` here.

From there, QpMatrixStuffing actually fills in the numeric matrices based on their symbolic representation, and the OSQP then makes small adjustments to match the exact inputs required for each solver, user-specified options, etc.

So let's look at what we hand over to OSQP!
For this you can call `data = problem.get_problem_data(cp.OSQP)`
Here is P, where I have added the variables as columns above for reference.
```
  t1, t2, x1, x2, y1, y2
[[2., 0., 0., 0., 0., 0.],
 [0., 2., 0., 0., 0., 0.],
 [0., 0., 0., 0., 0., 0.],
 [0., 0., 0., 0., 0., 0.],
 [0., 0., 0., 0., 0., 0.],
 [0., 0., 0., 0., 0., 0.]]
```
We see that the first two rows/columns refer to our new `t` variable (and the 2 offsets the 1/2 factor from the standard form)
Now let's look at our equality constraints $Ax=b$. 
We find that for A, we have
```
[[-1.,  0.,  1.,  0.,  1.,  0.],
 [ 0., -1.,  0.,  1.,  0.,  1.]]
```
and b is all zeros. That just means that `x + y == t`, great!
For our inequalities $Fx \leq G$, we have for F
```
[[ 0.,  0., -1.,  0.,  0.,  0.],
 [ 0.,  0.,  0., -1.,  0.,  0.],
 [ 0.,  0.,  0.,  0., -1.,  0.],
 [ 0.,  0.,  0.,  0.,  0., -1.]]
```
and G again all zeros. That means we have no inequality constraints on t, but enforce `x>=0` and `y>=0`.

That should basically be enough to get started digging into the transformations. Feel free to circle back if you have additional questions!
